crossbeam-epoch/src/lib.rs

```diff
@@ -61,20 +61,14 @@
 
 #[macro_use]
 extern crate cfg_if;
+#[cfg(feature = "alloc")]
+extern crate alloc;
 #[cfg(feature = "std")]
 extern crate core;
 
-cfg_if! {
-    if #[cfg(feature = "alloc")] {
-        extern crate alloc;
-    } else if #[cfg(feature = "std")] {
-        extern crate std as alloc;
-    }
-}
-
 #[cfg_attr(feature = "nightly", cfg(target_has_atomic = "ptr"))]
 cfg_if! {
-    if #[cfg(any(feature = "alloc", feature = "std"))] {
+    if #[cfg(feature = "alloc")] {
```

Comment:
- If `alloc` feature is on and `std` is off, then the compilation will fail, right?  I'd like to avoid such cases..
- I wonder if we can just remove the `alloc` feature and always require `alloc`.  Is it possible?

Reply:
> * If `alloc` feature is on and `std` is off, then the compilation will fail, right?  

No. Items in this `if` block are not require `std`. items that require `std` is in `if #[cfg(feature = "std")]` block: https://github.com/crossbeam-rs/crossbeam/blob/e782e91db8916b934eedb9714ef57ad08a619a0a/crossbeam-epoch/src/lib.rs#L92-L100

> I wonder if we can just remove the `alloc` feature and always require `alloc`. Is it possible?

Yes, the current crossbeam-epoch can do this, but I'd prefer to leave `alloc` feature for forward compatibility. (As long as `alloc` or `std` feature is not enabled, crossbeam-epoch with no-default-feature does not depend on `alloc` crate.)

Reply:
Oh sorry, I meant to ask what if `alloc` is off and `std` is on.  Then `mod default` will result in a compilation error, I guess.

Further, I don't understand the relationship between `alloc` and forward compatibility.  May I ask if you could further explain this point?

Reply:
@jeehoonkang 

> I meant to ask what if `alloc` is off and `std` is on. Then `mod default` will result in a compilation error, I guess.

Even if the user does not have `alloc` feature enabled, `alloc` feature is always enabled when `std` feature is enabled because [`std` feature depends on `alloc` feature](https://github.com/crossbeam-rs/crossbeam/pull/508/files#diff-a3fcd04172682b8bc7a8532346160690R21).

> I don't understand the relationship between `alloc` and forward compatibility. May I ask if you could further explain this point?

If we add a public API that does not depend on `alloc` to crossbeam-epoch, in order to use it in no-std environment where `alloc` crate cannot be used, it is necessary to separate APIs that require `alloc` crate by a feature gate.

However, this is a breaking change because adding this feature will prevent crate, which currently uses crossbeam-epoch with no-default-feature, from accessing APIs that require `alloc` crate.

To prevent this breaking change, from the beginning, APIs that require `alloc` crate should be under of a feature flags.

Reply:
Thank you for your kind explanation.  Now I can quite well understand your points.

One suggestion I'd like to make is that maybe we can write down a comment saying that `std` automatically imports `alloc`?  By doing so we can prevent someone like me from having difficulties why this should compile well always.

Reply:
> One suggestion I'd like to make is that maybe we can write down a comment saying that `std` automatically imports `alloc`? By doing so we can prevent someone like me from having difficulties why this should compile well always.

Sounds good, I'll add descriptions of the feature flag.